oarepo_vocabularies/services/cache.py

```diff
@@ -0,0 +1,209 @@
+        self.lru_terms_cache = TTLCache(maxsize=10000, ttl=3600)
+        self.count_from_cache = 0
+        self.count_fetched = 0
+        self.count_prefetched = 0
+
+    def _prefetch_vocabularies(self, vocabulary_types: List[str]):
+        yield from vocabulary_service.scan(
+            system_identity,
+            params={
+                "type": vocabulary_types,
+                "sort": "title",
+                "size": 1000,
+            },
+            preserve_order=True,
+        )
+
+    def _check_modified(self, cached_language, vocabulary_types: List[str]):
+        params = {
+            "size": 1,
+            "updated_after": {
+                vt: (
+                    cached_language[vt].last_modified.isoformat()
+                    if vt in cached_language
+                    else None
+                )
+                for vt in vocabulary_types
+            },
+        }
+
+        result = vocabulary_service.search_many(
+            system_identity,
+            params=params,
+        )
+        return result.total > 0
+
+    def _serialize_items(self, locale, items):
+        schema = VocabularyPrefetchSchema(context={"locale": locale})
+
+        for item in items:
+            dumped_item = schema.dump(item)
+            yield item, dumped_item
+
+    def get(self, vocabulary_types: List[str]):
+        self.update(vocabulary_types)
+        language = get_locale().language
+        ret = {}
+        for vocabulary_type in vocabulary_types:
+            ret[vocabulary_type] = self.cache[language][vocabulary_type].items
+            self.count_from_cache += len(ret[vocabulary_type])
+
+        return ret
+
+    def resolve(self, ids):
+        """
+        Resolves vocabulary ids to their localized records.
+
+        :param ids: list of vocabulary ids in the form of tuple (type, id)
+        """
+        locale = get_locale()
+        language = locale.language
+
+        # check if these are in the lru cache
+        cached, uncached, uncached_small = self._split_cached_uncached(language, ids)
+
+        self.count_from_cache += len(cached)
+        self.count_prefetched += len(uncached_small)
+        self.count_fetched += len(uncached)
+
+        if uncached_small:
+            self._fill_from_small_vocabularies_cache(language, uncached_small, cached)
+
+        if uncached:
+            for item, serialized_item in self._serialize_items(
+                locale,
+                vocabulary_service.search_many(
+                    system_identity, params={"ids": uncached}
+                ),
+            ):
+                typed_id = (item["type"], item["id"])
+                cached[typed_id] = serialized_item
+                self.lru_terms_cache[(language, typed_id)] = serialized_item
+
+        return cached
+
+    def _fill_from_small_vocabularies_cache(self, language, uncached_small, cached):
+        # TODO: is there a performance improvement for uncached_small or should we treat all as uncached?
```

Comment:
From my understanding a small cache is just a dict instance living inside the cache service instance...I cannot see much of a perf difference over uncached. Why don't we go either full cache (TTLCache) or non-cached here?

Reply:
The "small cache" are pre-fetched vocabularies for forms etc. It might be good to use those, because fetching facets from these means that forms will load faster (and vice versa). I do not have those numbers yet so I am not sure how big the speedup might be - thus the todo comment